src/modules/batch/tests_impl.h

```diff
@@ -0,0 +1,213 @@
+#ifndef SECP256K1_MODULE_BATCH_TESTS_H
+#define SECP256K1_MODULE_BATCH_TESTS_H
+
+#include "../../../include/secp256k1_batch.h"
+#ifdef ENABLE_MODULE_SCHNORRSIG
+#include "../../../include/secp256k1_schnorrsig.h"
+#include "../../../include/secp256k1_schnorrsig_batch.h"
+#endif
+#ifdef ENABLE_MODULE_EXTRAKEYS
+#include "../../../include/secp256k1_extrakeys.h"
+#include "../../../include/secp256k1_tweak_check_batch.h"
+#endif
+
+/* Tests for the equality of two sha256 structs. This function only produces a
+ * correct result if an integer multiple of 64 many bytes have been written
+ * into the hash functions. */
+void test_batch_sha256_eq(const secp256k1_sha256 *sha1, const secp256k1_sha256 *sha2) {
+    /* Is buffer fully consumed? */
+    CHECK((sha1->bytes & 0x3F) == 0);
+
+    CHECK(sha1->bytes == sha2->bytes);
+    CHECK(secp256k1_memcmp_var(sha1->s, sha2->s, sizeof(sha1->s)) == 0);
+}
+
+/* Checks that hash initialized by secp256k1_batch_sha256_tagged has the
+ * expected state. */
+void test_batch_sha256_tagged(void) {
+    unsigned char tag[13] = "BIP0340/batch";
+    secp256k1_sha256 sha;
+    secp256k1_sha256 sha_optimized;
+
+    secp256k1_sha256_initialize_tagged(&sha, (unsigned char *) tag, sizeof(tag));
+    secp256k1_batch_sha256_tagged(&sha_optimized);
+    test_batch_sha256_eq(&sha, &sha_optimized);
+}
+
+#define N_SIGS 10
+#define N_TWK_CHECKS 10
+#define N_TERMS (N_TWK_CHECKS + 2*N_SIGS)
+void test_batch_api(void) {
+
+#ifdef ENABLE_MODULE_EXTRAKEYS
+    unsigned char sk[32];
+    secp256k1_keypair keypair;
+    secp256k1_xonly_pubkey pk;
+    /* xonly pubkey tweak checks data */
+    unsigned char tweaked_pk[N_TWK_CHECKS][32];
+    int tweaked_pk_parity[N_TWK_CHECKS];
+    unsigned char tweak[N_TWK_CHECKS][32];
+    secp256k1_pubkey tmp_pk;
+    secp256k1_xonly_pubkey tmp_xonly_pk;
+    size_t i;
+#endif
+
+#ifdef ENABLE_MODULE_SCHNORRSIG
+    /* schnorr verification data */
+    unsigned char msg[N_SIGS][32];
+    unsigned char sig[N_SIGS][64];
+#endif
+    /* context and batch setup */
+    secp256k1_context *none = secp256k1_context_create(SECP256K1_CONTEXT_NONE);
+    secp256k1_context *sign = secp256k1_context_create(SECP256K1_CONTEXT_SIGN);
+    secp256k1_context *vrfy = secp256k1_context_create(SECP256K1_CONTEXT_VERIFY);
+    secp256k1_context *both = secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
+    secp256k1_context *sttc = secp256k1_context_clone(secp256k1_context_no_precomp);
+    secp256k1_batch *batch_none;
+    secp256k1_batch *batch_sign;
+    secp256k1_batch *batch_vrfy;
+    secp256k1_batch *batch_both;
+    secp256k1_batch *batch_sttc;
+    unsigned char aux_rand16[32];
```

Comment:
Getting another warning here when compiling together with Bitcoin Core:

```
src/modules/batch/tests_impl.h:68:19: warning: mixing declarations and code is a C99 extension [-Wdeclaration-after-statement]
    unsigned char aux_rand16[32];
                  ^
```

Reply:
Thanks @sipa, yeah, I should have added that I have this fixed now in my rebase branch: https://github.com/fjahr/secp256k1/commit/7c6b9df9ec8fe20fb228ead90d494e95f18c8344#diff-fa3fe2044a5385d080fb813c89c9d0707b50ec7083cef9cd9c114b1f2e5483c3R41